### PR TITLE
Support export to HuggingFace

### PR DESCRIPTION
**Pull request by pesvut**
Added option to export to huggingface format. This involves creating a folder with `config.json` and `pytorch_model.bin`.

Possible things that could be done: 
- Add export for tokenizer to huggingface format as well.
- Manually add the RoPE embeddings.
- Change versioning to be string (currently export to hf is when `version == -1`).
- Ensure that default values in LlamaConfig match up for the config files.

Loads without error into `LlamaForCausalLM.from_pretrained()` but have not tested with a tokenizer.

### CODE REVIEW

**Comment by pesvut** [comment]
Example of use:
```
curl -L -O https://huggingface.co/karpathy/tinyllamas/resolve/main/stories15M.pt
python export.py tinyllama-15m --checkpoint ./stories15M.pt --version -1              
```
I have also uploaded the models for use in [the HuggingFace Hub](https://huggingface.co/models?sort=trending&search=nickypro%2Ftinyllama). It seems to work fine, and automatically seems to get the RoPE embeddings when loading the model into LlamaForCausalLM.

File: export.py [hunk 1]
```diff
@@ -335,7 +335,7 @@ def permute_original(w, n_heads=llama_model.params.n_heads, dim1=llama_model.par
     os.makedirs(filepath, exist_ok=True)
 
     # Save the state dictionary in .pt format
```
Comment (xenova) [inline export.py:337]: ".pt" -> ".bin" in comments

Comment (xenova) [reply export.py:337]: Also, is there a way to choose dtype from the CLI? 👀

**Comment by xenova** [comment]
I've been testing this PR for the past ~hour and everything seems to work well! 🥳 

One thing which would be nice is the option to select dtype from the CLI.

**Comment by pesvut** [comment]
Ok, I changed the default dtype to be `torch.float32`. I also added code to choose dtype via CLI but it might not be worth merging upstream. I have made the fp16 and fp32 models [available on HuggingFace](https://huggingface.co/models?sort=trending&search=nickypro%2Ftinyllama)

**Comment by xenova** [comment]
This works perfectly, thanks!

```py
from transformers import pipeline
pipe = pipeline('text-generation', 'nickypro/tinyllama-110M')
pipe('Once upon a time, Sarah', max_new_tokens=128)
```
outputs
> [{'generated_text': "Once upon a time, Sarah and her mommy went to the park. Sarah saw a big slide and wanted to go on it. She climbed up the ladder and slid down. It was so much fun!\nAfter playing on the slide, Sarah and her mommy went to the store. Sarah saw a toy she really wanted, but her mommy said they couldn't buy it today. Sarah was sad, but her mommy promised they would come back another day.\nOn the way home, Sarah saw a big truck. It was carrying lots of boxes. Sarah asked her mommy what was inside the boxes. Her mommy"}]

**Comment by xenova** [comment]
Just one issue I ran into while converting to ONNX, is that the tied weights are not accounted for correctly. I ran this by @fxmarty, and he found that the state_dict is ill-formatted:
```py
model_id = "nickypro/tinyllama-15M"
model = AutoModelForCausalLM.from_pretrained(model_id)

print(model.model.embed_tokens.weight.data_ptr())
print(model.lm_head.weight.data_ptr())

##

model_id = "gpt2"
model = AutoModelForCausalLM.from_pretrained(model_id)

print(model.transformer.wte.weight.data_ptr())
print(model.lm_head.weight.data_ptr())
```

Perhaps Felix might be able to provide additional insight or a fix.

**Comment by fxmarty** [comment]
Hi, [this](https://huggingface.co/nickypro/tinyllama-15M/blob/main/config.json#L20) config needs to be set to `True` I believe.

**Comment by tic-top** [comment]
Have you tried to convert the [stories260K.pt](https://huggingface.co/karpathy/tinyllamas/tree/main/stories260K) to huggingface model?
It's architecture is GQA.

**Comment by pesvut** [comment]
> Hi, [this](https://huggingface.co/nickypro/tinyllama-15M/blob/main/config.json#L20) config needs to be set to `True` I believe.

Added check for tied weights, and fixed

**Comment by pesvut** [comment]
> Have you tried to convert the [stories260K.pt](https://huggingface.co/karpathy/tinyllamas/tree/main/stories260K) to huggingface model? It's architecture is GQA.

Added code to support GQA. It gives the same result for 15M, 42M etc, and runs for 260K. 

Not sure how to get the smaller tokenizer working with the HuggingFace model, so not able to check that it gives reasonable results.

**Comment by magician-blue** [comment]
> > Have you tried to convert the [stories260K.pt](https://huggingface.co/karpathy/tinyllamas/tree/main/stories260K) to huggingface model? It's architecture is GQA.
> 
> Added code to support GQA. It gives the same result for 15M, 42M etc, and runs for 260K.
> 
> Not sure how to get the smaller tokenizer working with the HuggingFace model, so not able to check that it gives reasonable results.

Good job!

**Comment by tic-top** [comment]
> > Have you tried to convert the [stories260K.pt](https://huggingface.co/karpathy/tinyllamas/tree/main/stories260K) to huggingface model? It's architecture is GQA.
> 
> Added code to support GQA. It gives the same result for 15M, 42M etc, and runs for 260K.
> 
> Not sure how to get the smaller tokenizer working with the HuggingFace model, so not able to check that it gives reasonable results.

good job

**Comment by KangkangStu** [comment]
> 不确定如何让较小的分词器与 HuggingFace 模型一起使用，因此无法检查它是否给出了合理的结果。

Hello，After trying to download a stories15M model on hf, I quantized it,
As shown in these two files。
image
then . I use " python export.py stories_q80.bin --version 2 --hf .\model\" That worked and it also succeeded for runq.c, but after executing” ./runq .stories_q80.bin -n 20 -i 'one day' “at the end, the code would just output my prompt one day and end the program
image

**Comment by SpaceCowboy850** [comment]
This thread is gold.  I'm posting this here in case someone wants a quick example using AutoModel instead of pipeline.

```
from transformers import AutoTokenizer, AutoModelForCausalLM
import os
import torch

torch.set_default_device("cuda")

model = AutoModelForCausalLM.from_pretrained(model_path, torch_dtype="auto", device_map="cuda", trust_remote_code=True)
tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)

prompt = "Once upon a time, Sarah"

tokenized_input = tokenizer(prompt, return_tensors="pt", return_attention_mask=False)

tokenized_output = model.generate(**tokenized_input, max_new_tokens=1024, repetition_penalty=1.1)
complete_response = tokenizer.batch_decode(tokenized_output)[0]
print(f"{complete_response}")
```

One thing I had to do to get this fully functional though was I pulled this:
https://huggingface.co/nickypro/tinyllama-110M/

And put in my own model that was exported from export.py with HF.  (config.json and pytorch_model.bin).

What I don't understand is how I'm supposed to properly create the other files in my huggingface folder export?
(e.g. generation_config.json, tokenizer.json, tokenizer_config.json, special_tokens_map.json)

When I look at meta/llama2 7B, the tokenizer file is not the same size as what is in nickypro/tinyllama-110M, nor is it the same size as what is in the root of llama2c.  So if I didn't have this "random" repo to set my tokenizer properly, I'm not clear how I would have done that.  And Nicky's is a json vs the 

In short - how would i recreate tokenizer.json?
